proposals/0110-exponential-fee-for-write-lock-accounts.md

```diff
@@ -0,0 +1,138 @@
+
+## New Terminology
+
+- *compute-unit utilization*: denominated in `cu`, it represents total
+ compute-units applied to a given resource.
+- *cost rate*: denominated in `lamport/cu`, it represents the cost per
+compute-unit at a given condition.
+- *compute unit pricer*: a componenet tracks Exponential Moving Average of
+*compute-unit utilization*, applies a pricing algorithm to provide current
+*cost rate*.
+- *write lock fee*: denominated in `lamport`, it is fee dedicated for write
+lock an account, calculated as `compute-unit-pricer.cost-rate() * transaction.requested_cu()`.
+
+## Design Highlights
+
+- Account Association with Compute Unit Pricer:
+  - Accounts are associated with a *compute unit pricer*, and the *runtime*
+  maintains an LRU cache of actively contentious accounts' public keys and
+  their *compute unit pricers*.
+  - Alternatively, each account can have its *compute unit pricer* stored
+  onchain, which would require modifying accounts.
+- Compute Unit Pricer:
+  - Tracks an account's EMA *compute-unit utilization*, updated after the
+  current bank is frozen.
+  - Provides the current *cost rate* when queried.
+- EMA of Compute-Unit Utilization:
+  - Uses 150 slots for EMA calculation.
+  - EMA Alpha, representing the degree of weighting decrease, is calculated as
+  `alpha = 2 / (N+1)`.
+- Pricing Algorithm:
+  - Adjusts write-lock *cost rate* based on an account's EMA *compute-unit
+  utilization*. Initial write-lock cost rate is `1000 lamport/CU`.
+  - For each block, if an account's EMA *compute-unit utilization* is more than
+  half of its max limit, its write-lock *cost rate* increases by 1%. If it's
+  below half, the *cost rate* decreases by 1%.
+- Calculate *Write Lock Fee*:
+  - Fee required to write-lock an account is calculated by multiplying the
+  write-lock *cost rate* by the transaction's requested CU.
+
+## Detailed Design
+
+- Initialization and Inheritance:
+  - Bank initializes an empty account_write_lock_fee_cache, an LRU Cache of
+  {account_pubkey, compute-unit-pricer}.
+  - Child banks inherit the parent's cache.
+- Transaction Fee Calculation:
+  - Calculate write-lock fee for each account a transaction needs to write,
+  summing up to be its *write lock fee*. This, along with signature fee and
+  priority fee, constitutes the total fee for the transaction.
+  - Leader checks fee payer's balance before scheduling the transaction.
+- Cost Tracking:
+  - Cost_tracker tracks CUs for the current block and each write-locked accounti
+  as-is;
+  - Ensuring cost tracking is enabled at the replay stage.
+- End of Block Processing:
+  - Identify write-locked accounts with *compute-unit utilization* > half of
+  account max CU limit. Add/update bank's account_write_lock_fee_cache. 
+  - Adding new account into LRU cache could push out eldest account;
+  - LRU cache has capacity of 1024, which should be large enough for hot accounts
+  in 150 slots.
```

Comment:
What is the relevance of 150 slots here?

Reply:
it's the # of slots until blockhash expiry, but also not sure why it's relevant here

Reply:
Attacker can saturate 128 * 8 accounts per block

Reply:
Max accounts in a tx is 128. Max txs with the accounts hitting the 6m CU limit is 8

Reply:
@taozhu-chicago 

i don't think 150 slot ema makes sense.  1.01x or 1/1.01x on every block depending if the account > 6m or < 6m, that will track the average already.  

Then the cache eviction policy is simple. 

Cache size = 2x * worst case number of accounts > 6m per block

Eviction = any new account > 6m CUs at the end of the block are added to the cache with a write lock fee of K. Cheapest accounts are evicted. 

Fee updates = if account is in the cache and > 6m fee is 1.01x, otherwise fee is 1/1.01x

Reply:
is the worry a malicious leader doing an attack to evict cache? if that's the concern, then we should be using a cache-size of 4096 since we allocate leader-slots in chunks of 4.

Reply:
@apfitzge second set woudl evict the first